google/cloud/storage/internal/openssl_util.h

```diff
@@ -0,0 +1,245 @@
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_OPENSSL_UTIL_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_OPENSSL_UTIL_H_
+
+#include "google/cloud/internal/throw_delegate.h"
+#include "google/cloud/storage/internal/credential_constants.h"
+#include "google/cloud/storage/version.h"
+#include <openssl/bio.h>
+#include <openssl/buffer.h>
+#include <openssl/evp.h>
+#include <openssl/opensslv.h>
+#include <openssl/pem.h>
+#include <algorithm>
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <vector>
+
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace internal {
+/**
+ * Helper functions for Base64 and related transcoding.
+ */
+struct OpenSslUtils {
+ public:
+  /**
+   * Encode a string using Base64.
+   */
+  static std::string Base64Encode(std::string const& str) {
+    auto bio_chain = MakeBioChainForBase64Transcoding();
+    int retval = 0;
+
+    while (true) {
+      retval = BIO_write(static_cast<BIO*>(bio_chain.get()), str.c_str(),
+                         static_cast<int>(str.length()));
+      if (retval > 0) break;  // Positive value == successful write.
+      if (not BIO_should_retry(static_cast<BIO*>(bio_chain.get()))) {
+        std::ostringstream err_builder;
+        err_builder << "Permanent error in " << __func__ << ": "
+                    << "BIO_write returned non-retryable value of " << retval;
+        google::cloud::internal::RaiseRuntimeError(err_builder.str());
+      }
+    }
+    // Tell the b64 encoder that we're done writing data, thus prompting it to
+    // add trailing '=' characters for padding if needed.
+    while (true) {
+      retval = BIO_flush(static_cast<BIO*>(bio_chain.get()));
+      if (retval > 0) break;  // Positive value == successful flush.
+      if (not BIO_should_retry(static_cast<BIO*>(bio_chain.get()))) {
+        std::ostringstream err_builder;
+        err_builder << "Permanent error in " << __func__ << ": "
+                    << "BIO_flush returned non-retryable value of " << retval;
+        google::cloud::internal::RaiseRuntimeError(err_builder.str());
+      }
+    }
+
+    // This buffer belongs to the BIO chain and is freed upon its destruction.
+    BUF_MEM* buf_mem;
+    BIO_get_mem_ptr(static_cast<BIO*>(bio_chain.get()), &buf_mem);
+    // Return a string copy of the buffer's bytes, as the buffer will be freed
+    // upon this method's exit.
+    return std::string(buf_mem->data, buf_mem->length);
+  }
+
+  /**
+   * Transform a string in-place, removing trailing occurrences of a character.
+   *
+   * Warning: this was written with the intent of operating on a string
+   * containing ASCII-encoded (8-bit) characters (e.g. removing trailing '='
+   * characters from a Base64-encoded string) and may not function correctly
+   * for strings containing Unicode characters.
+   */
+  static void RightTrim(std::string& str, char trim_ch) {
+    if (str.length() == 0) return;
```

Comment:
Consider `if (str.empty()) return;` instead.  Or better yet, consider removing that statement.  `find_last_not_of()` returns `npos` for an empty string, and the function will have no effect, and since the function is most often called with non-empty strings this statement is just slowing things down.

Reply:
Ah, good point.  Done (removed that line).